src/audio/smart_amp/smart_amp.c

```diff
@@ -0,0 +1,662 @@
+	if (sad->feedback_buf->source->state == dev->state) {
+		/* feedback */
+		avail_feedback_frames = sad->feedback_buf->stream.avail /
+			audio_stream_frame_bytes(&sad->feedback_buf->stream);
+
+		avail_frames = MIN(avail_passthrough_frames,
+				   avail_feedback_frames);
+
+		feedback_bytes = avail_frames *
+			audio_stream_frame_bytes(&sad->feedback_buf->stream);
+
+		buffer_unlock(sad->feedback_buf, feedback_flags);
+
+		comp_dbg(dev, "smart_amp_copy(): processing %d feedback frames (avail_passthrough_frames: %d)",
+			 avail_frames, avail_passthrough_frames);
+
+		sad->process(dev, &sad->feedback_buf->stream,
+			     &sad->sink_buf->stream, avail_frames,
+			     sad->config.feedback_ch_map, true);
+
+		comp_update_buffer_consume(sad->feedback_buf, feedback_bytes);
+	}
+
+	/* bytes calculation */
+	buffer_lock(sad->source_buf, &source_flags);
+	source_bytes = avail_frames *
+		audio_stream_frame_bytes(&sad->source_buf->stream);
+	buffer_unlock(sad->source_buf, source_flags);
+
+	buffer_lock(sad->sink_buf, &sink_flags);
+	sink_bytes = avail_frames *
+		audio_stream_frame_bytes(&sad->sink_buf->stream);
+	buffer_unlock(sad->sink_buf, sink_flags);
+
+	/* process data */
+	sad->process(dev, &sad->source_buf->stream, &sad->sink_buf->stream,
+		     avail_frames, sad->config.source_ch_map, false);
+
+	/* source/sink buffer pointers update */
+	comp_update_buffer_consume(sad->source_buf, source_bytes);
+	comp_update_buffer_produce(sad->sink_buf, sink_bytes);
+
+	return ret;
+}
+
+static int smart_amp_reset(struct comp_dev *dev)
+{
+	comp_dbg(dev, "smart_amp_reset()");
+
+	comp_set_state(dev, COMP_TRIGGER_RESET);
+
+	return 0;
+}
+
+static int smart_amp_prepare(struct comp_dev *dev)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	struct comp_buffer *source_buffer;
+	struct list_item *blist;
+	int ret;
+
+	comp_dbg(dev, "smart_amp_prepare()");
+
+	ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
+	if (ret < 0)
+		return ret;
+
+	/* searching for stream and feedback source buffers */
+	list_for_item(blist, &dev->bsource_list) {
+		source_buffer = container_of(blist, struct comp_buffer,
+					     sink_list);
+
+		if (source_buffer->source->comp.type == SOF_COMP_DEMUX)
+			sad->feedback_buf = source_buffer;
+		else
+			sad->source_buf = source_buffer;
+	}
+
+	sad->sink_buf = list_first_item(&dev->bsink_list, struct comp_buffer,
+					source_list);
+
+	sad->in_channels = sad->source_buf->stream.channels;
+	sad->out_channels = sad->sink_buf->stream.channels;
+
+	sad->feedback_buf->stream.channels = sad->config.feedback_channels;
+
+	/* TODO:
+	 * ATM feedback buffer frame_fmt is hardcoded to s32_le. It should be
+	 * removed when parameters negotiation between pipelines will prepared
+	 */
+	if (sad->feedback_buf->stream.frame_fmt != SOF_IPC_FRAME_S32_LE) {
+		sad->feedback_buf->stream.frame_fmt = SOF_IPC_FRAME_S32_LE;
+		comp_err(dev, "smart_amp_prepare(): S32_LE format is hardcoded as workaround");
```

Comment:
no return error code?

Reply:
@cujomalainey I considered to return error in smart_amp_process, but it causes several errors and blocks testing for capture stream. Hence I removed error return and added error message to move forward.

Reply:
Got it, as long as userspace handles the change properly then Im ok with this.